x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/AuthenticatorChain.java

```diff
@@ -0,0 +1,376 @@
+            getAuthenticatorConsumer(context, shouldExtractCredentials),
+            authenticators,
+            context.getThreadContext(),
+            Function.identity(),
+            result -> result.getStatus() == Authenticator.Status.UNSUCCESSFUL || result.getStatus() == Authenticator.Status.NOT_HANDLED);
+        iteratingActionListener.run();
+    }
+
+    private BiConsumer<Authenticator, ActionListener<Authenticator.Result>> getAuthenticatorConsumer(
+        Authenticator.Context context,
+        boolean shouldExtractCredentials
+    ) {
+        return (authenticator, listener) -> {
+            if (shouldExtractCredentials) {
+                final AuthenticationToken authenticationToken;
+                try {
+                    authenticationToken = authenticator.extractCredentials(context);
+                } catch (Exception e) {
+                    if (e instanceof ElasticsearchSecurityException) {
+                        listener.onFailure(e);
+                    } else { // other exceptions like illegal argument
+                        context.addUnsuccessfulMessage(authenticator.name() + ": " + e.getMessage());
+                        listener.onResponse(Authenticator.Result.unsuccessful(e.getMessage(), e));
+                    }
+                    return;
+                }
+                if (authenticationToken == null) {
+                    listener.onResponse(Authenticator.Result.notHandled());
+                    return;
+                }
+                context.addAuthenticationToken(authenticationToken);
+            }
+            context.setHandleNullToken(context.shouldHandleNullToken() && authenticator.canFollowedByNullTokenHandler());
+            authenticator.authenticate(context, ActionListener.wrap(result -> {
+                if (result.getStatus() == Authenticator.Status.UNSUCCESSFUL) {
+                    context.addUnsuccessfulMessage(authenticator.name() + ": " + result.getMessage());
+                }
+                listener.onResponse(result);
+            }, e -> {
+                if (e instanceof ElasticsearchSecurityException) {
+                    final ElasticsearchSecurityException ese = (ElasticsearchSecurityException) e;
+                    if (false == context.getUnsuccessfulMessages().isEmpty()) {
+                        addMetadata(context, ese);
+                    }
+                }
+                // Not adding additional metadata if the exception is not security related, e.g. server busy
+                listener.onFailure(e);
+            }));
+        };
+    }
+
+    private void maybeLookupRunAsUser(
+        Authenticator.Context context,
+        Authentication authentication,
+        ActionListener<Authentication> listener
+    ) {
+        // TODO: only allow run as for realm authentication to maintain the existing behaviour
+        if (false == runAsEnabled || authentication.getAuthenticationType() != Authentication.AuthenticationType.REALM) {
```

Comment:
Run-as is artificially restricted to realm authentication only to match existing behaviour.